### PR TITLE
Fix single_pass with built-in math functions

### PR DESCRIPTION
**Pull request by pawroman**
Using `single_pass` with built-in functions like `math.tanh` should be possible now.

### CODE REVIEW

File: fastats/core/decorator.py [hunk 1]
```diff
@@ -29,11 +29,7 @@ def fs(func):
 
     @wraps(func)
     def fs_wrapper(*args, **kwargs):
-        debug = kwargs.get('debug')
-        return_callable = kwargs.get('return_callable')
-
-        with suppress(KeyError):
-            del kwargs['return_callable']
```
Comment (pawroman) [inline fastats/core/decorator.py:36]: I see no reason for these shenanigans, did I miss anything?

File: fastats/core/decorator.py [hunk 2]
```diff
@@ -44,9 +40,16 @@ def fs_wrapper(*args, **kwargs):
...
             for v in kwargs.values():
-                if isfunction(v) and v.__name__ not in kwargs:
+                if v.__name__ in kwargs:
+                    continue
+
+                if isfunction(v):
                     new_funcs[v.__name__] = convert_to_jit(v)
-            kwargs = {k: convert_to_jit(v) for k, v in kwargs.items()}
+                elif isbuiltin(v):
+                    new_funcs[v.__name__] = v
```
Comment (pawroman) [inline fastats/core/decorator.py]: `math` functions are not regarded as functions by `isfunction`.

Comment (pawroman) [reply fastats/core/decorator.py]: Python 3.6.3 on Linux

```python
>>> import math
>>> import inspect
>>> inspect.isfunction(math.tanh)
False
>>> inspect.isbuiltin(math.tanh)
True
```

File: tests/core/ast_transforms/test_basic_sanity.py [hunk 3]
```diff
@@ -43,6 +54,35 @@ def test_child_transform_square_to_quadruple():
...
+def test_child_transform_with_faked_child():
+    # maliciously faking a function's name should not affect the result
+    # this can also happen when using decorators
+    assert child_faker.__name__ == child.__name__
+
+    original = parent(1)
+    assert original == 4
+
+    result = parent(1, child=child_faker)
+    assert result == 42
```
Comment (pawroman) [inline tests/core/ast_transforms/test_basic_sanity.py]: Fails here

File: fastats/core/decorator.py (hunk 2 above)
Comment (dwillmer) [reply fastats/core/decorator.py]: Yea, nice catch, thanks!

File: fastats/core/decorator.py (hunk 1 above)
Comment (dwillmer) [reply fastats/core/decorator.py:36]: Yea, we need to keep debug and return_callable until we have a better option. I liked your idea of having an optional context manager like `with substitute(x) as func:`, which would take care of most of the use cases of return_callable

File: tests/core/ast_transforms/test_basic_sanity.py (hunk 3 above)
Comment (dwillmer) [reply tests/core/ast_transforms/test_basic_sanity.py]: Yea, i don't think this is a valid use case - i'm happy to have this test assert an error if the `__name__` isn't the same as the function name. We are jit compiling, we have to have some restrictions :)

File: tests/optimize/root_finding/test_newton_raphson_hypo.py [hunk 4]
```diff
@@ -18,6 +18,7 @@ def less_or_equal(x, compared_to, rel=1e-6):
             or ((x - compared_to) == approx(0.0, rel=rel))
             or (x == approx(x, rel=rel)))
 
+
 nr_func = newton_raphson(1, 1e-6, root=func, return_callable=True)
```
Comment (dwillmer) [inline tests/optimize/root_finding/test_newton_raphson_hypo.py:22]: ...and these are the return_callable use-cases - for times when we only want to jit once, but have the function usable in multiple places (more 'framework' than 'exploratory data science').

File: tests/core/ast_transforms/test_basic_sanity.py (hunk 3 above)
Comment (pawroman) [reply tests/core/ast_transforms/test_basic_sanity.py]: As discussed, added xfail to mark as expected failure. Hopefully my next PR will fix this

File: fastats/core/decorator.py (hunk 1 above)
Comment (pawroman) [reply fastats/core/decorator.py:36]: OK - keeping it as is for now.

File: tests/core/ast_transforms/test_convert_to_jit.py [hunk 5]
```diff
@@ -16,23 +16,25 @@ def test_doesnt_convert_math_builtins():
 
 def test_doesnt_convert_jitted_functions():
     @jit
-    def jit_func():
+    def jit_func():     # pragma: no cover
         return 5
 
     assert convert_to_jit(jit_func) is jit_func
+    assert jit_func() == 5
```
Comment (pawroman) [inline tests/core/ast_transforms/test_convert_to_jit.py]: I don't like the pragma too much, but it strikes me as necessary evil. Some JIT code will simply be unreachable forever, and we need a mechanism to ignore such constructs.

**Comment by pawroman** [comment]
I'm happy with the code now - please let me know what you think of the `no cover` pragma

File: tests/core/ast_transforms/test_convert_to_jit.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+def test_does_not_convert_jitted_functions():
+    @jit
+    def jit_func():     # pragma: no cover
+        return 5
+
+    assert convert_to_jit(jit_func) is jit_func
+    assert jit_func() == 5
+
+
+def test_does_not_convert_converted_functions():
```
Comment (dwillmer) [inline tests/core/ast_transforms/test_convert_to_jit.py:26]: These are awesome, nice additions!

File: .coveragerc [hunk 7]
```diff
@@ -7,3 +7,4 @@ skip_covered = True
 exclude_lines =
     if __name__ == .__main__.:
     if self._debug:
+    pragma: no cover
```
Comment (dwillmer) [inline .coveragerc:10]: I have no issues with using  `no cover`, we've just tended to use trivial pure python assertions to give the test functions coverage (also means we have some validation for the test functions). Either way seems reasonable.

File: fastats/core/ast_transforms/convert_to_jit.py [hunk 8]
```diff
@@ -1,13 +1,25 @@
...
 def convert_to_jit(func):
-    if isinstance(func, CPUDispatcher):
+    if isinstance(func, CPUDispatcher) or isbuiltin(func):
         return func
 
-    _jit = jit(nopython=True, nogil=True)
+    if not isfunction(func):
+        raise TypeError("Can't JIT a non-function object: {}".format(func))
+
+    _jit = jit(**JIT_KWARGS)
```
Comment (dwillmer) [inline fastats/core/ast_transforms/convert_to_jit.py:21]: This is much neater, thanks!

File: tests/core/ast_transforms/test_basic_sanity.py [hunk 9]
```diff
@@ -43,6 +56,46 @@ def test_child_transform_square_to_quadruple():
...
+    final_two = parent(2)
+    assert final_two == 16
+
+    final = parent(3)
+    assert final == 36
+
+
+def test_problematic_child_transform_with_faked_child():
+    # function with a faked name should be respected as an override
+    # TODO this test captures current problematic behaviour which should be fixed
```
Comment (dwillmer) [inline tests/core/ast_transforms/test_basic_sanity.py]: I'm still not convinced by this - faking a function name is always going to mess up a jit compiler, so i don't see this as a valid use case. I think we should just remove this test.

Comment (pawroman) [reply tests/core/ast_transforms/test_basic_sanity.py]: I think we should document the library's shortcomings. Tests are good way of capturing this - "you don't have to take our word for it, just take a look at the test". This also encourages contribution, to address "known issues".

File: .coveragerc (hunk 7 above)
Comment (pawroman) [reply .coveragerc:10]: That's fine, but to test the functions that do the actual jitting (just like in this module), we need this sort of mechanism or the coverage thing is going to pester us

File: tests/core/ast_transforms/test_basic_sanity.py [hunk 10]
```diff
@@ -73,20 +73,23 @@ def test_child_transform_square_to_zero():
...
 
 def test_problematic_child_transform_with_faked_child():
-    # function with a faked name should be respected as an override
-    # TODO this test captures current problematic behaviour which should be fixed
+    # Because of the intricacies of the AST rewrite
+    # (global namespace per transformed function), it's hard
+    # to make sure we're resilient to function name faking.
+    #
+    # This is a known shortcoming of our approach
+    # - you're welcome to fix it!
```
Comment (pawroman) [inline tests/core/ast_transforms/test_basic_sanity.py:81]: Changed the comment as agreed. I still believe it's possible to fix it - never say never!

Comment (pawroman) [reply tests/core/ast_transforms/test_basic_sanity.py:81]: Should be good to merge now.

**Comment by dwillmer** [comment]
Awesome stuff, @pawroman!

**Pull request merged**